contrib/lib-autopilot.py

```diff
@@ -0,0 +1,289 @@
+            return False
+        self.__logger.info("Number of nodes found: {}".format(len(nodes)))
+
+        for node in nodes:
+            self.G.add_node(node["nodeid"],**node)
+        return True
+    
+    def __load_edges(self):
+        """
+        tries to store channels of the lightning network to the networkx graph. the attributes of the
+        channel will be stored as edge attributes
+        """
+        if len(self.G.nodes)==0:
+            self.__add_logger("cannot download channels if nodes do not exist. Try downloading the node list...")
+            if self.__load_nodes() == False:
+                self.add_logger("Stop trying to download channels")
+                return False
+        
+        channels = {}
+        try: 
+            self.__logger.info("Attempt RPC-call to download channels from the lightning network")
+            channels = self.__rpc_interface.listchannels()["channels"]
+            self.__logger.info("Number of retrieved channels: {}".format(len(channels)))
+        except ValueError as e:
+            self.__logger.info("Channel list could not be retrieved from the peers of the lightning network")
+            self.__logger.debug("RPC error: " + str(e))
+            return False
+        
+        for channel in channels:
+            self.G.add_edge(channel["source"],channel["destination"],**channel)
+        
+        return True
+
+    def __generate_rich_nodes(self,k=3,percentile=0.5):
+        """
+            generates a set of nodes that are very well connected (such nodes with low degree will be prefered)
+        """
+        if k < 3:
+            k = 3
+        
+        if percentile < 0.0 or percentile > 1.0:
+            percentile = 0.5
+        
+        self.__logger.info("RICH_NODES: Try to generate {} candidates from the {}-percentile".format(k,percentile))
+        
+        rich_nodes = {}
+        network_capacity = 0
+        candidates = []
+        for n in self.G.nodes():
+            total_capacity = sum(self.G.get_edge_data(n,m)["satoshis"] for m in self.G.neighbors(n))
+            network_capacity += total_capacity
+            rich_nodes[n] = total_capacity
+            
+        cumsum = 0
+        for n,value in sorted(rich_nodes.items(),key=itemgetter(1),reverse=True):
+            cumsum +=value
+            self.__logger.debug("RICH_NODES: node {} has {} neighbors and a balance of {}".format(n,value,len(list(self.G.neighbors(n)))))
+            candidates.append(n)
+            if cumsum > network_capacity*percentile:
+                break 
+        
+        self.__logger.info("RICH_NODES: Found {} candidates in the {}-percentile".format(len(candidates),percentile))        
+        if len(candidates) <= k:
+            return list(candidates)
+        self.__logger.info("RICH_NODES: Sample {} items from the candidates as was requested".format(k))
+        tmp = list(candidates)
+        random.shuffle(tmp)
+        return tmp[0:k]
```

Comment:
No need to `random.shuffle`, you can use `random.sample()` to do the same (without actually having to touch all elements).

Reply:
switched to sampling from `numpy` as I refactored everything to use probability distributions